### PR TITLE
Feature/ai 161 improve data filtering

### PR DESCRIPTION
**Pull request by cschaefer26**
- More sophisticated filtering of the dataset according to duration stats
- Filtering cleanup

### CODE REVIEW

**Review by alexteua — APPROVED** [review]
Nice functionality!

File: utils/dataset.py [hunk 1]
```diff
@@ -171,23 +207,105 @@ def __len__(self) -> int:
...
+                         num_workers=0) -> Tuple[DataLoader, DataLoader]:
+
+    """Returns training and validation dataloaders.
+
+    Args:
+        paths (Paths): An instance of the Paths class containing file paths.
+        batch_size (int): The batch size for the dataloaders.
+        max_mel_len (int, optional): The maximum length of the mel spectrograms. Defaults to 1250.
+        filter_duration_stats (bool): Whether to filter accordint to below duration stats.
+        min_attention_alignment (float): The minimum attention alignment value. Defaults to 0.5.
```
Comment (alexteua) [inline utils/dataset.py:284]: I think more organic would be to set those default values in the function definition itself

File: utils/dataset.py [hunk 2]
```diff
@@ -282,96 +412,65 @@ def get_binned_taco_dataloader(paths: Paths, max_batch_size: int = 8) -> BinnedT
...
+        speaker_counts_orig = Counter([speaker_dict[item_id] for item_id, _ in train_data + val_data if item_id in speaker_dict])
+        train_data = data_filter(train_data)
+        val_data = data_filter(val_data)
+        speaker_counts_filtered = Counter([speaker_dict[item_id] for item_id, _ in train_data + val_data if item_id in speaker_dict])
 
-    def __init__(self, taco_collator: TacoCollator) -> None:
-        self.taco_collator = taco_collator
+        table = [(speaker, count, speaker_counts_orig[speaker]-speaker_counts_filtered[speaker])
+                 for speaker, count in speaker_counts_filtered.most_common()]
+        print(tabulate(table, headers=('speaker', 'files used', 'files removed')))
```
Comment (alexteua) [inline utils/dataset.py:445]: additional column of total entries would make it more intuitive

File: utils/dataset.py [hunk 3]
```diff
@@ -282,96 +412,65 @@ def get_binned_taco_dataloader(paths: Paths, max_batch_size: int = 8) -> BinnedT
...
+        duration_stats = unpickle_binary(paths.duration_stats)
+        data_filter = DataFilter(duration_stats=duration_stats,
+                                 min_attention_alignment=min_attention_alignment,
+                                 min_attention_sharpness=min_attention_sharpness,
+                                 max_consecutive_duration_ones=max_consecutive_ones,
+                                 max_duration=max_duration)
 
-class ForwardCollator:
+        speaker_counts_orig = Counter([speaker_dict[item_id] for item_id, _ in train_data + val_data if item_id in speaker_dict])
+        train_data = data_filter(train_data)
```
Comment (alexteua) [inline utils/dataset.py:439]: While filtering train and val subsets there is a possible situation when distribution will be changed, for example, for certain speakers samples will be completely gone from train and kept in val or smth like that. But then the whole flow should be reviewed

File: utils/dataset.py (hunk 1 above)
Comment (cschaefer26) [reply utils/dataset.py:284]: Removed the defaults from docstring. There is actually no default, so values should be taken from the config.

File: utils/dataset.py (hunk 3 above)
Comment (cschaefer26) [reply utils/dataset.py:439]: Could happen. I'd suggest to tackle this in a separate PR if needed.

**Review by alexteua — APPROVED** [review]

**Pull request merged**